= trapdiff
Paul Klemm <paul.klemm@sf.mpg.de>
:toc:
:repo: https://github.com/paulklemm/trapdiff
:releases: {repo}/releases/tag
:issues: {repo}/issues
:docs: https://pages.github.com/paulklemm/trapdiff
:images: images

IMPORTANT: This package is work in progress. It is not suited for use in production.

== Overview

The goal of this package is to have an automated interface for conducting RNA-seq analysis of riboTRAP and bacTRAP data.

In this data, we have two measurements per sample: Input (the background) and IP (pulldown of a specific cell type). We call this *source*.
Then we have a experimental *treatment*, for example normal chow diet (NCD) and high fat diet (HFD).

We are now interested in genes that are (1) enriched in our pulldown (IP) and (2) differentially expressed between our experimental treatment (e.g. HFD vs NCD) in the IP.

We do this by using link:https://bioconductor.org/packages/release/bioc/html/DESeq2.html[DESeq2] with an interaction term model that looks as follows:

```r
design = source + treatment + source:treatment
```

== Usage

We define our experiment using a JSON file.

.Here is an example.
[%collapsible]
====

```json
{
  "source": {
    "cd": [
      "cd1_ip_gfp",
      "cd2_ip_gfp",
      "cd3_ip_gfp",
      "cd4_ip_gfp",
      "cd1_input",
      "cd2_input",
      "cd3_input",
      "cd4_input"
    ],
    "hfd": [
      "hfd1_ip_gfp",
      "hfd2_ip_gfp",
      "hfd3_ip_gfp",
      "hfd4_ip_gfp",
      "hfd1_input",
      "hfd2_input",
      "hfd3_input",
      "hfd4_input"
    ]
  },
  "treatment": {
    "ip": [
      "cd1_ip_gfp",
      "cd2_ip_gfp",
      "cd3_ip_gfp",
      "cd4_ip_gfp",
      "hfd1_ip_gfp",
      "hfd2_ip_gfp",
      "hfd3_ip_gfp",
      "hfd4_ip_gfp"
    ],
    "input": [
      "cd1_input",
      "cd2_input",
      "cd3_input",
      "cd4_input",
      "hfd1_input",
      "hfd2_input",
      "hfd3_input",
      "hfd4_input"
    ]
  },
  "treatment_comparison": [
    "ip",
    "input"
  ],
  "source_comparison": [
    "hfd",
    "cd"
  ]
}
```

====

In `treatment_comparison` we say which comparison we want.
In this example, we compare `IP vs Input`.

You also need to provide a counts object that contains a tabular representation of your count matrix. 
It must not be normalized to library size (see DESeq2 requirements on input count matrices).
It needs to contain a column `ensembl_gene_id` and a column for each sample.

*The counts are rounded to integers to comply with DESeq2 standards during the analysis!*

Additionally, we require a TPM object that contains library-size normalized counts.
It must contain a `ensembl_gene_id` column.

== Shiny

TrapDiff also provides a shiny app that allows for interactively analyzing the trapdiff output.
Find the app with `inst/rmd/trapdiff_shiny/app.R`. You can send the path via the URL, e.g. `http://127.0.0.1:6252/?trappath=/beegfs/scratch/bruening_scratch/pklemm/2020-05-yiyi-bactrap/release/trapdiff`

== Development

These options are only important for developing the package.

.Dev
[%collapsible]
====

Shiny-App Dev-version
[source,r]
----
shiny::runApp("inst/rmd/trapdiff_shiny")
# Open with /?trappath=/beegfs/scratch/bruening_scratch/pklemm/2020-05-yiyi-bactrap/release/trapdiff to open a dataset
----

Debug the RMarkdown doc.

Anna's data

[source,r]
----
mywd <- getwd()
counts_data <- "/beegfs/scratch/bruening_scratch/pklemm/2017-10-anna7-ribotrap/nfcore-rnaseq-analysis/results/salmon/salmon_merged_gene_counts.csv" %>%
tpms_data <- "/beegfs/scratch/bruening_scratch/pklemm/2017-10-anna7-ribotrap/nfcore-rnaseq-analysis/results/salmon/salmon_merged_gene_tpms.csv" %>%
  readr::read_csv() %>%
  dplyr::rename(ensembl_gene_id = gene_id)
xaringan::infinite_moon_reader(
  moon = "inst/rmd/trapdiff/trapdiff.Rmd",
  cast_from = file.path(mywd, "inst", "rmd", "trapdiff"),
  params = list(
    path_config_json = paste0(mywd, "/trapdiff-debug/config_il6.json"),
    biotypes_filter = "protein_coding",
    debug_cache = TRUE,
    counts_min = 30,
    save_rds = TRUE,
    save_excel = TRUE,
    counts = counts_data,
    tpms = tpms_data
  )
)
----

Yiyi's data

[source,r]
----
mywd <- getwd()
counts_data <- "/beegfs/scratch/bruening_scratch/pklemm/2020-05-yiyi-bactrap/nfcore-rnaseq-pipeline/results/salmon/salmon_merged_gene_counts.csv" %>%
tpms_data <- "/beegfs/scratch/bruening_scratch/pklemm/2020-05-yiyi-bactrap/nfcore-rnaseq-pipeline/results/salmon/salmon_merged_gene_tpms.csv" %>%
  readr::read_csv() %>%
  dplyr::rename(ensembl_gene_id = gene_id)

xaringan::infinite_moon_reader(
  moon = "inst/rmd/trapdiff/trapdiff.Rmd",
  cast_from = file.path(mywd, "inst", "rmd", "trapdiff"),
  params = list(
    path_config_json = paste0(mywd, "/trapdiff-debug/config_yiyi.json"),
    biotypes_filter = "protein_coding",
    counts_min = 30,
    save_rds = TRUE,
    save_excel = TRUE,
    counts = counts_data,
    counts = tpms_data
  )
)
----

Alex' bacTRAP data

[source,r]
----
mywd <- getwd()
counts_data <- "/beegfs/scratch/bruening_scratch/pklemm/2019-02-alex-trap-nfcore/nfcore-rnaseq-pipeline/l10a/results/salmon/salmon_merged_gene_counts.csv" %>%
tpms_data <- "/beegfs/scratch/bruening_scratch/pklemm/2019-02-alex-trap-nfcore/nfcore-rnaseq-pipeline/l10a/results/salmon/salmon_merged_gene_tpms.csv" %>%
  readr::read_csv() %>%
  dplyr::rename(ensembl_gene_id = gene_id)

xaringan::infinite_moon_reader(
  moon = "inst/rmd/trapdiff/trapdiff.Rmd",
  cast_from = file.path(mywd, "inst", "rmd", "trapdiff"),
  params = list(
    path_config_json = paste0(mywd, "/trapdiff-debug/config_alex.json"),
    biotypes_filter = "protein_coding",
    counts_min = 30,
    save_rds = TRUE,
    save_excel = TRUE,
    counts = counts_data
    tpms = tpms_data
  )
)
----

====
