= trapdiff
Paul Klemm <paul.klemm@sf.mpg.de>
:toc:
:repo: https://github.com/paulklemm/trapdiff
:releases: {repo}/releases/tag
:issues: {repo}/issues
:docs: https://pages.github.com/paulklemm/trapdiff
:images: images

== Overview

The goal of this package is to have an automated interface for conducting RNA-seq analysis of riboTRAP and bacTRAP data.

In this data, we have two measurements per sample: Input (the background) and IP (pulldown of a specific cell type). We call this *source*.
Then we have a experimental *condition*, for example normal chow diet (NCD) and high fat diet (HFD).

We are now interested in genes that are (1) enriched in our pulldown (IP) and (2) differentially expressed between our experimental condition (e.g. HFD vs NCD) in the IP.

We do this by using link:https://bioconductor.org/packages/release/bioc/html/DESeq2.html[DESeq2] with an interaction term model that looks as follows:

```r
design = source + condition + source:condition
```

== Usage

We define our experiment using a JSON file.
Here is an example.

```json

{
  "ip": ["ip_hfd_1", "ip_hfd_3", "ip_hfd_4", "ip_hfd_2", "ip_ncd_3", "ip_ncd_2", "ip_ncd_1"],
  "input": ["input_hfd_4", "input_hfd_1", "input_hfd_2", "input_hfd_3", "input_ncd_2", "input_ncd_1", "input_ncd_3"],
  "condition": {
    "hfd": ["wt_ip_hfd_1", "wt_input_hfd_4", "wt_ip_hfd_3", "wt_input_hfd_1", "wt_ip_hfd_4", "wt_input_hfd_2", "wt_input_hfd_3", "wt_ip_hfd_2"],
    "ncd": ["wt_input_ncd_2", "wt_input_ncd_1", "wt_ip_ncd_3", "wt_ip_ncd_2", "wt_ip_ncd_1", "wt_input_ncd_3"]
  },
  "condition_comparison": ["hfd", "ncd"]
}

```

In `condition_comparison` we say which comparison we want.
In this example, we compare `HFD vs NCD`.

== History

.History
[%collapsible]
====
* _2020-06-26_
** Started project
====