= trapdiff
Paul Klemm <paul.klemm@sf.mpg.de>
:toc:
:repo: https://github.com/paulklemm/trapdiff
:releases: {repo}/releases/tag
:issues: {repo}/issues
:docs: https://pages.github.com/paulklemm/trapdiff
:images: images

IMPORTANT: This package is work in progress. It is not suited for use in production.

== Overview

The goal of this package is to have an automated interface for conducting RNA-seq analysis of riboTRAP and bacTRAP data.

In this data, we have two measurements per sample: Input (the background) and IP (pulldown of a specific cell type). We call this *source*.
Then we have a experimental *treatment*, for example normal chow diet (NCD) and high fat diet (HFD).

We are now interested in genes that are (1) enriched in our pulldown (IP) and (2) differentially expressed between our experimental treatment (e.g. HFD vs NCD) in the IP.

We do this by using link:https://bioconductor.org/packages/release/bioc/html/DESeq2.html[DESeq2] with an interaction term model that looks as follows:

```r
design = source + treatment + source:treatment
```

== Usage

We define our experiment using a JSON file.

.Here is an example.
[%collapsible]
====

```json
{
  "source": {
    "ip": [
      "wt_ip_ncd_1",
      "wt_ip_ncd_2",
      "wt_ip_ncd_3",
      "wt_ip_hfd_1",
      "wt_ip_hfd_2",
      "wt_ip_hfd_3",
      "wt_ip_hfd_4"
    ],
    "input": [
      "wt_input_ncd_1",
      "wt_input_ncd_2",
      "wt_input_ncd_3",
      "wt_input_hfd_1",
      "wt_input_hfd_2",
      "wt_input_hfd_3",
      "wt_input_hfd_4"
    ]
  },
  "treatment": {
    "hfd": [
      "wt_ip_hfd_1",
      "wt_ip_hfd_2",
      "wt_ip_hfd_3",
      "wt_ip_hfd_4",
      "wt_input_hfd_1",
      "wt_input_hfd_2",
      "wt_input_hfd_3",
      "wt_input_hfd_4"
    ],
    "ncd": [
      "wt_ip_ncd_1",
      "wt_ip_ncd_3",
      "wt_ip_ncd_2",
      "wt_input_ncd_1",
      "wt_input_ncd_2",
      "wt_input_ncd_3"
    ]
  },
  "treatment_comparison": [
    "hfd",
    "ncd"
  ]
}
```

====

In `treatment_comparison` we say which comparison we want.
In this example, we compare `HFD vs NCD`.

You also need to provide a counts object that contains a tabular representation of your count matrix.
It needs to contain a column `ensembl_gene_id` and a column for each sample.

*The counts are rounded to integers to comply with DESeq2 standards during the analysis!*

== Development

These options are only important for developing the package.

.Dev
[%collapsible]
====
Debug the RMarkdown doc.

[source,r]
----
mywd <- getwd()
xaringan::infinite_moon_reader(
  moon = "inst/rmd/trapdiff/trapdiff.Rmd",
  cast_from = file.path(mywd, "inst", "rmd", "trapdiff"),
  params = list(
    path_config_json = paste0(mywd, "/trapdiff-debug/config.json"),
    biotypes_filter = "protein_coding",
    debug_cache = TRUE,
    counts_min = 30,
    counts = trapdiff::get_test_counts() %>%
      # Add empty columns to check if detection of these work
      dplyr::mutate(
        debug_sample_1 = 1,
        debug_sample_2 = 2
      )
  )
)
----

====

== History

.History
[%collapsible]
====
* _2020-06-26_
** Started project
====